File: Data Structures & Algorithms/coin-change/submission-6.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = {}
        def dfs(rem):
            if rem == 0:
                return 0
            if rem < 0:
                return -1
            if rem in dp:
                return dp[rem]
            minRes = float('inf')
            for i in range(len(coins)):
                res = dfs(rem-coins[i])

                if res >= 0:
                    minRes = min(res+1,minRes)
            dp[rem] = -1 if minRes == float('inf') else minRes
            return dp[rem]
        return dfs(amount)
            

File: Data Structures & Algorithms/min-cost-climbing-stairs/submission-6.cpp
class Solution {
public:
  vector<int> cache;
  int minCostClimbingStairs(vector<int> &cost) {
    cache.resize(cost.size(), -1);
    return min(dfs(0, cost), dfs(1, cost));
  }
  int dfs(int i, const vector<int> &cost) {
    if (i >= cost.size()) {
      return 0;
    }
    if (cache[i] != -1) {
      return cache[i];
    }
    return cache[i] = cost[i] + min(dfs(i + 1, cost), dfs(i + 2, cost));
  }


};
